File: shared-spark-session-helper/src/main/scala/de/example/spark/testing/job/AwesomeJob.scala
// Author: Gustavo Martin Morcuende
package de.example.spark.testing.job

import com.typesafe.scalalogging.LazyLogging
import de.example.spark.testing.job.AwesomeJob.{Database, Table}
import de.example.spark.testing.service.AwesomeService
import org.apache.spark.sql.{SaveMode, SparkSession}
import org.apache.spark.sql.types.{StringType, StructField, StructType}

private object AwesomeJob {
  private val Database = "testing"
  private val Table = "example"
}

class AwesomeJob(sourcePath: String, destinationPath: String, awesomeService: AwesomeService)(
    implicit sparkSession: SparkSession)
    extends LazyLogging {

  def run(): Unit = {
    logger.info("Running AwesomeJob")

    val jsonSchema = StructType(
      Array(
        StructField("name", StringType),
        StructField("surname", StringType)
      )
    )
    val dataFrame = sparkSession.read.schema(jsonSchema).json(sourcePath)
    val schema = dataFrame.schema

    val upperCaseSchema = awesomeService.renameColumnsToUpperCase(schema)
    val upperCaseDataFrame = sparkSession.createDataFrame(dataFrame.rdd, upperCaseSchema)

    sparkSession.sql(s"CREATE DATABASE IF NOT EXISTS $Database")
    sparkSession.sql(s"""
                        |CREATE TABLE IF NOT EXISTS `$Database`.`$Table` (`NAME` STRING, `SURNAME` STRING)
                        |USING PARQUET
                        |OPTIONS (
                        |  path '$destinationPath'
                        |)
                        |""".stripMargin)
    upperCaseDataFrame.write
      .mode(SaveMode.Overwrite)
      .insertInto(s"$Database.$Table")

    logger.info("End running AwesomeJob")
  }
}

File: pyspark/example/awesome_job.py
class AwesomeJob:

    def __init__(self, source_path, destination_path, spark_session, awesome_service):
        self.source_path = source_path
        self.destination_path = destination_path
        self.spark_session = spark_session
        self.awesome_service = awesome_service

    def run(self):
        json_data_frame = self.spark_session.read.json(self.source_path)
        json_schema = json_data_frame.schema
        upper_case_json_schema = self.awesome_service.rename_columns_to_upper_case(json_schema)
        upper_case_json_data_frame = self.spark_session.createDataFrame(json_data_frame.rdd, upper_case_json_schema)
        upper_case_json_data_frame.write.json(self.destination_path)
